### ISSUE TITLE
Performance of gen_hard_sample

### ISSUE THREAD

**Issue by Chuniba** [CongWeilin/mtcnn-caffe#21]
Im trying to genarate the hard sample with gen_12net_hard_example.py. it seems take much time. 
Did I get something wrong?

**Comment by CongWeilin** [comment]
yes it take a lot of time because 12net generate many images, especially on wider face( many face on 1 image)

**Comment by knsong** [comment]
hi @CongWeilin , when generating hard examples for 24net using `gen_12net_hard_example.py`,  I got ~1000 negative examples for one positve example. It seems that I will get at least ~60 million negative examples on wider face.  Is it normal?

**Comment by wuzhiyang2016** [comment]
@Trunibal  hei ,i have trained pnet and draw detection reslults on test pictures, too many boxes than @CongWeilin 's model,,,i think my train is something wrong ,,could you show me your train log? need your help~

**Comment by CongWeilin** [comment]
@wuzhiyang2016 check out your training image preprocess code. Might something wrong. OR, train cls model as an initiate for pts and roi training is a better option.

**Comment by wuzhiyang2016** [comment]
@CongWeilin  thanks very much, the training image preprocess code is follow your gen_net_imdb.py  
i only train classification and box regression these two tasks.


import numpy as np
import numpy.random as npr
size = 12
net = str(size)

with open('E:/MtcnnCaffe/mtcnn-caffe-master/prepare_data/pnet/test_imdb/pos_%s.txt' % (net), 'r') as f:
    pos = f.readlines()

with open('E:/MtcnnCaffe/mtcnn-caffe-master/prepare_data/pnet/test_imdb/neg_%s.txt' % (net), 'r') as f:
    neg = f.readlines()

with open('E:/MtcnnCaffe/mtcnn-caffe-master/prepare_data/pnet/test_imdb/part_%s.txt' % (net), 'r') as f:
    part = f.readlines()


def view_bar(num, total):  
    rate = float(num) / total
    rate_num = int(rate * 100) + 1
    r = '\r[%s%s]%d%%' % ("#" * rate_num, " " * (100 - rate_num), rate_num,)
    sys.stdout.write(r)
    sys.stdout.flush()


import sys
import cv2
import os
import numpy as np

cls_list = []
print '\n' + 'positive-12'
cur_ = 0
sum_ = len(pos) 
for line in pos:
    view_bar(cur_, sum_)
    cur_ += 1
    words = line.split()  # 12/positive/0 1 0.17 0.08 -0.03 0.05
    image_file_name = 'E:/mtcnn-master/data/' + words[0] + '.jpg'
    # print image_file_name
    im = cv2.imread(image_file_name)
    h, w, ch = im.shape
    # print h
    if h != 12 or w != 12:
        im = cv2.resize(im, (12, 12))
    im = np.swapaxes(im, 0, 2)
    im = (im - 127.5) / 127.5  
    label = 1
    roi = [-1, -1, -1, -1]
    pts = [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1]  
    cls_list.append([im, label, roi])
print '\n' + 'negative-12'
cur_ = 0
neg_keep = npr.choice(len(neg), size=600000, replace=False)
sum_ = len(neg_keep)
for i in neg_keep:
    line = neg[i]
    view_bar(cur_, sum_)
    cur_ += 1
    words = line.split()
    image_file_name = 'E:/mtcnn-master/data/' + words[0] + '.jpg'
    im = cv2.imread(image_file_name)

    h, w, ch = im.shape
    if h != 12 or w != 12:
        im = cv2.resize(im, (12, 12))
    im = np.swapaxes(im, 0, 2)
    im = (im - 127.5) / 127.5
    label = 0
    roi = [-1, -1, -1, -1]
    pts = [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
    cls_list.append([im, label, roi])
import cPickle as pickle

fid = open("E:/MtcnnCaffe/mtcnn-caffe-master/prepare_data/pnet/test_imdb/cls_128.imdb", 'w')
pickle.dump(cls_list, fid)
fid.close()

roi_list = []
print '\n' + 'part-12'
cur_ = 0

part_keep = npr.choice(len(part), size=300000, replace=False)
sum_ = len(part_keep)
for i in part_keep:
    line = part[i]
    view_bar(cur_, sum_)
    cur_ += 1
    words = line.split()
    image_file_name = 'E:/mtcnn-master/data/' + words[0] + '.jpg'
    im = cv2.imread(image_file_name)
    h, w, ch = im.shape
    if h != 12 or w != 12:
        im = cv2.resize(im, (12, 12))
    im = np.swapaxes(im, 0, 2)
    im = (im - 127.5) / 127.5  # im -= 128 
    label = -1
    roi = [float(words[2]), float(words[3]), float(words[4]), float(words[5])]
    pts = [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
    roi_list.append([im, label, roi])
print '\n' + 'positive-12'
cur_ = 0
sum_ = len(pos)
for line in pos:
    view_bar(cur_, sum_)
    cur_ += 1
    words = line.split()
    image_file_name = 'E:/mtcnn-master/data/' + words[0] + '.jpg'
    im = cv2.imread(image_file_name)
    h, w, ch = im.shape
    if h != 12 or w != 12:
        im = cv2.resize(im, (12, 12))
    im = np.swapaxes(im, 0, 2)
    im = (im - 127.5) / 127.5
    label = -1
    roi = [float(words[2]), float(words[3]), float(words[4]), float(words[5])]
    pts = [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
    roi_list.append([im, label, roi])
import cPickle as pickle

fid = open("E:/MtcnnCaffe/mtcnn-caffe-master/prepare_data/pnet/test_imdb/roi_128.imdb", 'w')
pickle.dump(roi_list, fid)
fid.close()
print "finished all data!!!"

**Comment by wuzhiyang2016** [comment]
@CongWeilin  so  i am so puzzled , cant't find what wrong with me~

**Comment by wuzhiyang2016** [comment]
@CongWeilin   i go to do  **train cls model as an initiate for pts and roi training is a better option**   for a try

**Comment by CongWeilin** [comment]
@wuzhiyang2016. The acc should be about 94% i can not remember the loss.

**Comment by wuzhiyang2016** [comment]
@CongWeilin  ok i did't set test dataset,,  i will prepare one

**Comment by walkerwjt** [comment]
@wuzhiyang2016 hi, i have the same quesion with you, i have trained the PNet by using 12net-cls-only.caffemodel as an initial model, but when i generate 12net hard examples, i still get too many boxes than @CongWeilin ' model. My training acc is 89%-95%, sometimes, it can get 100%, so i am really puzzed. Have you solved your problem? i need your help.

**Comment by AITTSMD** [comment]
@wuzhiyang2016  I have the same quesion with u. Pnet and Rnet I trained are bad.Training acc is about 90%,but when I use them to generate hard examples,I can't get many positive examples.
